crates/rethnet_evm/src/blockchain/storage.rs

```diff
@@ -0,0 +1,25 @@
+mod contiguous;
+mod sparse;
+
+use rethnet_eth::{B256, U256};
+
+pub use self::{contiguous::ContiguousBlockchainStorage, sparse::SparseBlockchainStorage};
+
+/// An error that occurs when trying to insert a block into storage.
+#[derive(Debug, thiserror::Error)]
+pub enum InsertError {
+    /// Block already exists
+    #[error("A block with hash {block_hash} and number {block_number} already exists.")]
+    DuplicateBlock {
+        /// Duplicated block hash
+        block_hash: B256,
+        /// Duplicated block number
+        block_number: U256,
+    },
```

Comment:
If a hash is duplicated, the number **must** be duplicated too, given that the hash calculation includes the number.

Theoretically, it could be that a block with with a duplicated number has a different hash, but it feels like overkill to have two errors: `DuplicateBlockDueToHash` and `DuplicateBlockDueToNumber`. The error is the fact that the block is duplicated. I will however change the member variable comments to not say "Duplicated" to avoid confusion.

Reply:
Resolved by https://github.com/NomicFoundation/hardhat/pull/4175/commits/bf48a23f39ef5be9e021426a7b1f1a2c0e910aa0

Reply:
Addressed in https://github.com/NomicFoundation/hardhat/pull/4175/commits/aa7be38e1e37acac4b8fe035c7d4ef3b17cd8d4b by adding punctuation

From a linguistic perspective, it doesn't actually state the the number AND hash are duplicated. The block already exists and these are its hash and number.